src/Authentication/Actions/CreateIdentityTrait.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+namespace CodeIgniter\Shield\Authentication\Actions;
+
+use CodeIgniter\Shield\Entities\User;
+use CodeIgniter\Shield\Models\DatabaseException;
+use CodeIgniter\Shield\Models\UserIdentityModel;
+
+/**
+ * Create an identity for auth action
+ */
+trait CreateIdentityTrait
+{
+    /**
+     * Create an identity for auth action
+     */
+    private function createIdentity(User $user, string $name, string $extra): string
+    {
+        helper('text');
+
+        /** @var UserIdentityModel $userIdentityModel */
+        $userIdentityModel = model(UserIdentityModel::class);
+
+        // Delete any previous identities for action
+        $userIdentityModel->deleteIdentitiesByType($user, $this->type);
+
+        // Create an identity for the action
+        $maxTry = 5;
+        $data   = [
+            'user_id' => $user->getAuthId(),
+            'type'    => $this->type,
+            'name'    => $name,
+            'extra'   => $extra,
+        ];
+
+        while (true) {
+            $data['secret'] = $this->generateSecretCode();
+
+            try {
+                $userIdentityModel->create($data);
+
+                break;
+            } catch (DatabaseException $e) {
```

Comment:
Instead of trying to insert the record and catching an exception, would it be simpler to just check if an identity of that type with that code exists in the database? No need to worry about what types of exceptions are thrown then.

Reply:
Even If we check that the code exists in the database, we still need to catch the exception.
Because it is a race condition.

Process A: generate code "123456"
Process B: generate code "123456"
Process A: read the database --> OK
Process B: read the database --> OK
Process A: insert --> OK
Process B: insert --> Error

Alternatively, the table needs to be locked before reading, but CI4 has no way to do that.
And locking does not seem a good way.

Reply:
Good point.

Reply:
I understand the race condition, but realistically what we are coding around is generating two random six-digits strings that are the same within milliseconds of each other. The theoretical scope of unlikelihood has to be on the magnitude of trillions. I'm fine with simplifying this to a database read to clear up the DBDebug stuff - we could add a comment in the code about race condition to remember that it was not overlooked.